### ISSUE TITLE
add lint for having a Panics comment when code cannot panic

### ISSUE THREAD

**Issue by dylanopen** [rust-lang/rust-clippy#15327]
### What it does

There exists a Clippy lint, `missing_panics_doc`, which fails if you have a function which panics but you haven't documented why. In other words, this code would fail:

``` rust
/// this function does stuff and this is a doc comment
/// there's no panics section here
fn this_may_panic() {
    if (...) {
        panic!("oops");
    }
}
```

Clippy will warn you about this if you have `clippy::missing_panics_doc` or `clippy::pedantic` enabled. However, there doesn't seem to be a lint for the other way round.

Clippy considers this code fine, even though the doc comment is misleading:

```rust
/// This function does stuff
/// 
/// ## Panics
/// 
/// - This function will panic if something happens (it actually won't)
fn get_number() -> i32 {
    return 23;
}
```

> It may also be a good idea to do this with the `Errors` section. Doing this for an out-of-date `Errors` section (i.e. the function no longer returns a `Result` or `Option` but still has an `Errors` section) would be much easier to implement as a lint, because Clippy would only need to look at the function signature.
> I understand if implementing this for panics would be too difficult, but I think it can be done quite easily for `Error` comments on functions that don't return an error.

### Advantage

- It would be confusing for (library) users to see a `Panics` section if it is impossible for the function to panic
- I realised this would be a good lint when I nearly committed to git a change, without updating the doc comment to specify that panics were now impossible.
- It's easy for comments to go outdated, so Clippy should try to help keep comments up to date as much as possible.

### Drawbacks

It is possible for some Rust constructs to panic even if the callstack doesn't contain the `panic!` macro, so this lint could be a little tricky to implement.

If the function uses the `as` keyword, for example, the program may crash (in debug mode) even though none of the functions or macros invoked actually used the `panic!` macro. So it's possible that code that seems not to panic, could still panic.

### Example

```rust
/// This function does stuff
/// 
/// ## Panics
/// 
/// - This function will panic if something happens (it actually won't)
fn get_number() -> i32 {
    return 23;
}
```

Should instead be written as:

```rust
/// This function does stuff (no panics section)
fn get_number() -> i32 {
    return 23;
}
```

**Comment by dylanopen** [comment]
Not in any way important or urgent, but I thought I'd bump this given I posted it ~ a year ago :)
